### PR TITLE
Root Makefile and devcontainer

### PR DESCRIPTION
**Pull request by 4ydan**
## The Makefile:
Here's a Markdown table for each command:

| Command | Description |
| --- | --- |
| `run-frontend` | Builds the frontend and runs the development server |
| `run-backend` | Builds the backend and runs the application |
| `test` | Runs tests for the frontend, backend, and mdbook |
| `test-frontend` | Installs dependencies, checks formatting, lints, and runs tests for the frontend |
| `test-backend` | Runs tests for the backend |
| `test-mdbook` | Runs tests for mdbook |
| `storybook` | Runs Storybook for the frontend |
| `build` | Builds bindings, frontend, backend, Storybook, and mdbook |
| `build-frontend` | Builds bindings, installs dependencies, generates API types, and builds the frontend |
| `build-backend` | Builds bindings, sets up diesel, runs migrations, and builds the backend |
| `build-storybook` | Installs dependencies and builds Storybook for the frontend |
| `build-mdbook` | Installs mdbook and mdbook plugins, and builds mdbook |
| `bindings` | Generates TypeScript bindings for the backend |
| `psql-r` | Connects to the PostgreSQL database remotely |
| `pre-commit-all` | Runs pre-commit hooks on all files |
| `clean` | Cleans frontend and backend directories |
| `clean-frontend` | Removes frontend dependencies |
| `clean-backend` | Cleans the backend and mdbook directories |

These commands would be used in the CI pipeline.

**Should I maybe write three Makefiles, one in the root, one in the backend and one in the frontend folder?**

## Devcontainers
Since devcontainers seem to be an upcoming topic (also with github Codespaces) and my way of developing, I have created this environment. Also this is kinda the byproduct of the base image for the CI containers.

### Benefits
- Attractive for developers ready to go environment with a database connection and pgadmin
- Can clone/open pull request into the containers
- Already installed pre-commit for everyone developing inside the container

### Downside
- Might be slower than developing on the host
- Building the image for the first time takes a bit of time
- Something additional that has to be "maintained" but removing it isn't really "breaking" anything

### Try it out
If you meet the prerequites for devcontainers you can try it out with this branch by opening vscode and cloning the pull request.
<kbd>F1</kbd> `Dev Containers: Clone Github Pull Request in Container Volume...` and paste the URL of this pullrequest.

## Basics

- [ ] I added a line to [/doc/CHANGELOG.md](/doc/CHANGELOG.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.

## Checklist

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Comment by markus2330** [comment]
> Here's a Markdown table for each command:

Please add this to our README.md

> .devcontainer

Please add it to /doc/setups. Does it maybe replace a setup from there?

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	mdbook build
+
+# MISC
+
+.PHONY: bindings
+bindings:
+	typeshare ./backend --lang=typescript --output-file=./frontend/src/bindings/definitions.ts
+
+.PHONY: psql-r
+psql-r:
```
Comment (markus2330) [inline Makefile:63]: Very nice extras! :heart:

**Review by markus2330 — COMMENTED** [review]
Btw. Makefile and devcontainer should be better be separate PRs.

File: Makefile (hunk 1 above)
Comment (4ydan) [reply Makefile:63]: Thank you.

**Pull request closed without merging**